tests/Basics/SupervisedBufferTest.cpp

```diff
@@ -434,46 +452,207 @@ TEST(SupervisedBufferTest, MultipleGrowsAndRecycle) {
     int growths = 0;
     for (int i = 0; i < 512 && growths < 2; ++i) {
+      std::size_t capBefore = supervisedBuffer.capacity();
+      std::size_t monitorBefore = monitor.current();
+
       builder.add(Value("abcdefghijklmnopqrstuvwxyz"));
-      if (supervisedBuffer.capacity() > lastCapacity) {
-        ASSERT_GT(monitor.current(), lastMonitorCurrent);  // monitor increased
-        ASSERT_GE(monitor.current(), builder.size());
-        lastCapacity = supervisedBuffer.capacity();
-        lastMonitorCurrent = monitor.current();
+
+      std::size_t capAfter = supervisedBuffer.capacity();
+      std::size_t monitorAfter = monitor.current();
+
+      ASSERT_GE(monitorAfter, builder.bufferRef().size());
+
+      if (capAfter > capBefore) {
+        std::size_t capDiff = capAfter - capBefore;
+        ASSERT_EQ(monitorAfter, monitorBefore + capDiff);
         ++growths;
-      } else {
-        ASSERT_GE(monitor.current(), builder.size());
       }
     }
     ASSERT_GE(growths, 2);
+    ASSERT_GE(growths, 2);
     builder.close();
     ASSERT_GE(monitor.current(), builder.size());
 
-    // Recycle the buffer, memory doesn't decrease
+    // recycle the buffer, memory doesn't decrease
     auto maintainedMonitorCurrent = monitor.current();
     builder.clear();
     ASSERT_EQ(monitor.current(), maintainedMonitorCurrent);
 
-    // Append items but capacity must remain the same
+    // capacity must remain the same
     builder.openArray();
     for (int i = 0; i < 16; ++i) {
       builder.add(Value("abcde"));
-      ASSERT_GE(monitor.current(), builder.size());
+      ASSERT_GE(monitor.current(), builder.bufferRef().size());
       ASSERT_EQ(monitor.current(), maintainedMonitorCurrent);
     }
     builder.close();
     ASSERT_EQ(monitor.current(), maintainedMonitorCurrent);
   }
 
   ASSERT_EQ(monitor.current(), 0);
-}
\ No newline at end of file
+}
+
+TEST(SupervisedBufferTest, StealWithMemoryAccounting) {
+  auto& global = GlobalResourceMonitor::instance();
+  ResourceMonitor monitor{global};
+
+  {
+    // scope that will own the bytes after the steal
+    ResourceUsageScope owningScope(monitor);
+
+    SupervisedBuffer supervisedBuffer(monitor);
+    Builder builder(supervisedBuffer);
+
+    // the local inline capacity is 192 bytes
+    std::size_t localCap = supervisedBuffer.capacity();
+
+    // force growth beyond local capacity
+    builder.openArray();
+    for (int i = 0; i < 64; ++i) {
+      builder.add(Value(std::string(256, 'a')));
+    }
+    builder.close();
+
+    // buffer grew, so it allocated
+    ASSERT_GT(supervisedBuffer.capacity(), localCap);
+
+    // total and buffer capacity immediately before stealing
+    std::size_t beforeTotal = monitor.current();
+    std::size_t capBefore = supervisedBuffer.capacity();
+    ASSERT_EQ(beforeTotal, capBefore);
+
+    uint8_t* stolen = supervisedBuffer.stealWithMemoryAccounting(owningScope);
+    // get the builder's ptr back to start
+    builder.clear();
```

Comment:
Holy, moly

Let me understand this.
You are stealing the underlying buffer data here.
But the builder is probably still thinking it has something in it. 🤔 

**Nothing to fix here.**
But is this actually a dangerous combination?

Reply:
I think it's safe here at least in the test because noone else is using the builder in between stealWithMemoryAccounting() and clear(), and builder.close() is called before stealing